marlowe-runtime/src/Language/Marlowe/Protocol/Sync/Codec.hs

```diff
@@ -0,0 +1,134 @@
+{-# LANGUAGE GADTs #-}
+
+module Language.Marlowe.Protocol.Sync.Codec where
+
+import Data.Binary (get, getWord8, put, putWord8)
+import qualified Data.ByteString.Lazy as LBS
+import Language.Marlowe.Protocol.Sync.Types
+import Language.Marlowe.Runtime.ChainSync.Api (BlockHeader)
+import Language.Marlowe.Runtime.Core.Api (MarloweVersion (..), SomeMarloweVersion (..))
+import Network.Protocol.Codec (DeserializeError, GetMessage, PutMessage, binaryCodec)
+import Network.TypedProtocol.Codec (Codec, PeerHasAgency (..), SomeMessage (SomeMessage))
+
+codecMarloweSync :: Applicative m => Codec MarloweSync DeserializeError m LBS.ByteString
+codecMarloweSync = binaryCodec putMessage getMessage
+  where
+  putMessage :: PutMessage MarloweSync
+  putMessage = \case
+    ClientAgency TokInit -> \case
+      MsgFollowContract contractId -> do
+        putWord8 0x01
+        put contractId
+      MsgIntersect contractId version blockHeaders -> do
+        putWord8 0x02
+        put contractId
+        put $ SomeMarloweVersion version
+        put blockHeaders
+
+    ServerAgency TokFollow -> \case
+      MsgContractNotFound -> putWord8 0x03
+      MsgContractFound blockHeader version createStep -> do
+        putWord8 0x04
+        put blockHeader
+        put $ SomeMarloweVersion version
+        case version of
+          MarloweV1 -> put createStep
+
+    ClientAgency (TokIdle _) -> \case
+      MsgRequestNext -> putWord8 0x05
+      MsgDone        -> putWord8 0x06
+
+    ServerAgency (TokNext version) -> \case
+      MsgRollForward blockHeader contractSteps -> do
+        putWord8 0x07
+        put blockHeader
+        case version of
+          MarloweV1 -> put contractSteps
+      MsgRollBackward blockHeader -> do
+        putWord8 0x08
+        put blockHeader
+      MsgRollBackCreation -> putWord8 0x0e
+
+      MsgWait -> putWord8 0x09
+
+    ClientAgency (TokWait _) -> \case
+      MsgPoll   -> putWord8 0x0a
+      MsgCancel -> putWord8 0x0b
+
+    ServerAgency (TokIntersect _) -> \case
+      MsgIntersectFound blockHeader -> do
+        putWord8 0x0c
+        put blockHeader
+      MsgIntersectNotFound -> putWord8 0x0d
+
+  getMessage :: GetMessage MarloweSync
+  getMessage tok = do
+    tag <- getWord8
+    case tag of
+      0x01 -> case tok of
```

Comment:
Are these purely internal intra-Runtime tags, or do we need to document these for external (non-Runtime) use?

Reply:
No, this protocol is an external-facing API, and these tag values are part of the binary serialization schema for the messages. So yes, this does need to be documented. Haskell clients will be able to use the typed interface which already implements this serialization scheme, but implementations in other languages will need to explicitly follow the specification.